airflow/api_fastapi/core_api/datamodels/log.py

```diff
@@ -16,11 +16,29 @@
 # under the License.
 from __future__ import annotations
 
-from pydantic import BaseModel
+from datetime import datetime
+from typing import Annotated
+
+from pydantic import BaseModel, ConfigDict, WithJsonSchema
+
+
+class StructuredLogMessage(BaseModel):
+    """An individual log message."""
+
+    # Not every message has a timestamp.
+    timestamp: Annotated[
+        datetime | None,
+        # Schema level, say this is always a datetime if it exists
+        WithJsonSchema({"type": "string", "format": "date-time"}),
```

Comment:
I'm not sure the `WithJsonSchema` is necessary. This is automatically generated when the type is `datetime` in the final openapi spec file.

Reply:
This is to not include the `None` option as a possible value to the client. At the python level this is either a datetime, or one. But when we serialize this, if it's present it's always a datetime.

So the override is needed so that the JS doesn't see the properly as possibly nullable. A subtle distinction but a distinction none the less.

Reply:
like the key could be missing from the object ? Is that the use case ?

Yes unfortunately we do not have an easy way to specify that. Maybe we need two distinct models…

Anyway we needed that at other places and we chose not to manually override the generated json. 

I’m not strong on that but I would recommend we  don’t do it here as well. At least for consistency,  also because. null check on the front end is a one liner anyway. And finally because the code and the spec could get out of sync by manually doing that. What happens if the code actually return 200 ? Does that raises a 422 or return a response not in line with the spec ?

Reply:
This `WithJsonSchema` makes the schema an accurate representation of what we actually send is why.

`{"timestamp": "2025-02-24T14:04:03"}`
`{}`

are both allowed but

`{"timestamp": null}`

is not.

The risk of it getting out of sync is minimal -- it's on the next line. Yes, if the code tried to send a None it would return an error.

> Anyway we needed that at other places and we chose not to manually override the generated json.

Such as? Why did you choose not to override it?